Create spotlight on analytics revenue report

octaedro opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

After adding the [date sorting configuration for Analytics](https://github.com/woocommerce/woocommerce/pull/36492), the payment date is assigned to the order creation date, regardless of when the payment was made. This approach can be misleading and lead to complications in terms of taxes and invoices. We'll change the default date used in this report and introduce the flexibility to use different dates: created, paid or completed.

When the user visits the Revenue and the Orders report for the first time (after we implement the change) we will show a spotlight with more context about it


![Screenshot 2023-01-27 at 15 16 19](https://user-images.githubusercontent.com/1314156/215163595-7002506b-1669-4283-bdb5-a0987d59c64e.png)


![Screenshot 2023-02-21 at 16 20 45](https://user-images.githubusercontent.com/1314156/220438956-89f82b61-4e67-46d4-89ab-f16f5597f06e.png)


Closes #35887.

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

1. Checkout this branch.
2. Go to Analytics > Revenue.
3. Verify that the spotlight is being shown.
4. Press `Got it` or the `x`.
5. Refresh the page.
6. Verify that the spotlight is not visible anymore.
7. Do the same steps above but for Analytics > Orders. (Spotlight should appear again)

If you want to repeat the test, you should delete the `woocommerce_date_type`, `woocommerce_revenue_report_date_tour_shown` and `woocommerce_orders_report_date_tour_shown ` options. 
The tour should not be shown if `woocommerce_date_type` is already defined.

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

octaedro committed:
Add `RevenueReportDateTour` component

octaedro committed:
Add changelog

octaedro committed:
Fix styles

louwie17 reviewed (changes requested):
This tested well, but this shouldn't be merged until: https://github.com/woocommerce/woocommerce/pull/36492 has been merged.
That's why I kept it as request changes for now.

nathanss committed:
Rename RevenueReportDateTour to ReportDateTour to use it in two different reports: Orders and Revenue

nathanss committed:
Change default sort date to date_paid

# Conflicts:
#	plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php

nathanss committed:
Update changelog

octaedro commented:
Hey @louwie17,
I just rebased this PR due to a conflict with `trunk`. Since https://github.com/woocommerce/woocommerce/pull/36492 has been merged, could you take another look at this PR?

@nathanss added a couple of commits to make the `report-date-tour` component more flexible and reusable for orders.

octaedro committed:
Merge branch 'trunk' into add/35887_spotlight_on_analytics_revenue

octaedro committed:
Merge branch 'trunk' into add/35887_spotlight_on_analytics_revenue

nathanss committed:
Merge branch 'trunk' into add/35887_spotlight_on_analytics_revenue

nathanss committed:
Assign date_column_name before calling parent constructor

louwie17 commented on plugins/woocommerce-admin/client/guided-tours/report-date-tour.tsx:
```diff
@@ -0,0 +1,94 @@
+			select( OPTIONS_STORE_NAME );
+
+		return {
+			hasShownTour: getOption( optionName ) === 'yes',
```
Would it also be worth checking if the `woocommerce_date_type` is actually set to `date_paid` or maybe still `undefined` (meaning it has not been set yet)?

Mostly wondering as @nathanss PR that adds the `woocommerce_date_type` option has already been merged before the code freeze. So this PR won't make it in until the next release (unless we make a code freeze exception).
This could mean that users have already changed that option themselves and were already aware of that.

nathanss replied:
I would leave it like this. I might be wrong but I don't expect many users to change this option before the guided tour, since it's kind of hidden in the analytics settings. @octaedro what do you think?

octaedro replied:
Maybe we can confirm this with @pmcpinto or @AnnaMag 
Should we try to include this PR in 7.5?

AnnaMag replied:
@octaedro it would be helpful to include this alongside other change, but I'm afraid it does not qualify for a code freeze exception. Do you think differently? 

If we aim for 7.6 with this change, I'd agree with @louwie17's suggestion to be on a safe side as we have seen cases of users manipulating hidden and apparently low-traffic options.

louwie17 replied:
Yeah I am afraid it won't qualify for a code freeze exception either.
The change should be relatively straight forward, so I don't think it will hurt adding it in, even if the impact will be low.

nathanss replied:
I will implement this today

louwie17 reviewed (commented):
This tested well, nice work @octaedro. I did leave one question in the code, depending on the outcome of that question I will approve :)

nathanss committed:
Remove date_type default value from configuration and show tour only if user didn't assign a value

louwie17 commented on plugins/woocommerce-admin/client/analytics/settings/config.js:
```diff
@@ -143,6 +148,5 @@ export const config = applyFilters( SETTINGS_FILTER, {
 			'Database date field considered for Revenue and Orders reports',
 			'woocommerce'
 		),
-		defaultValue: 'date_created',
```
We probably want to keep this and set it to `date_paid` it does feel slightly strange going to the settings screen and seeing the `Select a date type`.
But it did look like the `defaultValue` doesn't work right now, so it would have to be added here: https://github.com/woocommerce/woocommerce/blob/4c821a6b4edc889e99abce857eada3c0bcc4b5f8/plugins/woocommerce-admin/client/analytics/settings/setting.js#L82 (`value || defaultValue`)

nathanss replied:
I removed it because I was having problems when clicking on save without touching the date field. In the case where the option was not defined, the default value would be shown, leading the user to believe that the option is set to the default value. However, when clicking on save without changing the value (touching the field), the value is not persisted (the field is not perceived as "dirty")... That's why I opted to not use defaultValue for now and set a value for the option only when the user explicitly defines it. (Even though the behavior for the reports will default to date_paid when the option is not defined).

louwie17 reviewed (approved):
Nice work @nathanss this tested well, thanks for making the adjustment as well!

I did leave one comment, but I think this PR is good as is, and that one could be a follow up, given its a slight bug in the settings field anyway that `defaultValue` is not being used.

nathanss committed:
Merge branch 'trunk' into add/35887_spotlight_on_analytics_revenue
